src/GafferOSL/ShadingEngine.cpp

```diff
@@ -191,6 +195,21 @@ class RenderState
 
 		bool userData( ustring name, TypeDesc type, void *value ) const
 		{
+			if( name == gIndex )
+			{
+				// if a 4 byte type has been requested then ensure we fit and cast to narrower type
+				// this way f32 reads of _index will succeed.
+				if( type.size() == sizeof( int ) && m_pointIndex <= ( (size_t) std::numeric_limits<int>::max() ) )
+				{
+					int v = ( int ) m_pointIndex;
+					return ShadingSystem::convert_value( value, type, &v, OIIO::TypeDesc( OIIO::TypeDesc::INT32 ) );
+				}
+				else
+				{
+					return ShadingSystem::convert_value( value, type, &m_pointIndex, OIIO::TypeDesc( OIIO::TypeDesc::UINT64 ) );
```

Comment:
I committed a version which emits a warning but it looks out of place as we don't do that in the other cases where we return false from RenderState::userData. I don't have strong opinions either way on this.